docs/proposals/07-shoot-control-plane-migration.md

```diff
@@ -0,0 +1,244 @@
+          "modules": [
+              {
+              }
+          ]
+          ...
+      }
+```
+
+Extensions which do not require state migration should set `status.state=nil` in their Custom Resources and trigger a normal reconciliation operation if the CR contains the `core.gardener.cloud/operation=restore` annotation.
+
+Similar to the contract for the [reconcile operation](https://github.com/gardener/gardener/blob/master/docs/extensions/reconcile-trigger.md), the extension controller has to remove the `restore` annotation after the restoration operation has finished.
+
+An additional annotation `gardener.cloud/operation=migrate` is added to the Custom Resources. It is used to tell the extension controllers in the __Source Seed__ that they must stop reconciling resources (in case they are requeued due to errors) and should perform cleanup activities in the Shoot's control plane. These cleanup activities involve removing the finalizers on Custom Resources and deleting them without actually deleting any infrastructure resources.
+
+**Note:** The same size limitations from the previous section are relevant here as well.
+
+### Shoot reconciliation flow changes
+
+The only data which must be stored in the `ShootSnapshot` by the gardenlet is secrets (e.g ca for the API server). Therefore the `botanist.DeploySecrets` step is changed. It is split into two functions which take a list of secrets that have to be generated.
+- `botanist.GenerateSecretState` Generates certificate authorities and other secrets which have to be persisted in the ShootSnapshot and must not be regenerated on the __Destination Seed__.
+- `botanist.DeploySecrets` Takes secret data from the `ShootSnapshot`, generates new ones (e.g. client tls certificates from the saved certificate authorities) and deploys everything in the Shoot's control plane on the __Destination Seed__
+
+
+### ShootSnapshot synchronization controller
+
+The ShootSnapshot synchronization controller will become part of the gardenlet. It syncs the state of extension custom resources from the shoot namespace to the garden cluster and updates the corresponding `spec.extension.state` field in the `ShootSnapshot` resource. The controller can `watch` Custom Resources used by the extensions and update the `ShootSnapshot` only when changes occur.
+
+
+### Migration workflow
+1. Starting migration
+    - Migration can only be started after a Shoot cluster has been successfully created so that the `status.seed` field in the `Shoot` resource has been set
+    - The `Shoot` resource's field `spec.seedName="new-seed"` is edited to hold the name of the __Destination Seed__ and reconciliation is automatically triggered
+    - The Garden Controller Manager checks if the equality between `spec.seedName` and `status.seed`, detects that they are different and triggers migration.
+1. The Garden Controller Manager waits for the __Destination Seed__ to be ready
+1. Shoot's API server is stopped
+1. Backup the Shoot's ETCD.
```

Comment:
IIUC, several shoots can be backed up in the same bucket. And there is one bucket per seed to store backups of all shoots that belong to that seed. Is it correct?

After a shoot is migrated, is it going to be backed up in a different bucket? I wonder what would happen if a shoot is first migrated from SeedA to SeedB, and then migrated back to SeedA. Would it be possible then for the shoot to reuse the outdated backup in SeedA's bucket instead of the more recent backup from SeedB's bucket?

Does the Garden Controller Manager know which bucket to use by reading the field `myshoot.status.seed`, which is guaranteed to points to the most recent seed after a completed migration?

I am wondering what would happen if the Garden Controller Manager inconveniently crashes just before this "Backup"  step.

Reply:
> Who is responsible to update this field to the Destination Seed, and when is it updated?

There is #2012 - it introduces a flow for the shoot migration. Basically, the control plane in the old/source seed is teared down, and as last step the `BackupEntry`'s `.spec.seedName` is changed to the new/destination seed, https://github.com/gardener/gardener/pull/2012/files#diff-94a8ce02ab81b8c9d1db83c32e6b8b55R252

> IIUC, several shoots can be backed up in the same bucket. And there is one bucket per seed to store backups of all shoots that belong to that seed. Is it correct?

Correct.

> After a shoot is migrated, is it going to be backed up in a different bucket?

No, it will keep using the same/old bucket. 

> Would it be possible then for the shoot to reuse the outdated backup in SeedA's bucket instead of the more recent backup from SeedB's bucket?

Yes, it's possible, we are not able to move the data from bucket 1 to bucket 2. Hence, it's kept where it is. cc @swapnilgm 

> Does the Garden Controller Manager know which bucket to use

The trigger for the above mentioned migration preparation flow is to change `.spec.seedName=<destination-seed>` in the `Shoot`. After this flow is finished the `BackupEntry`'s `.spec.seedName` is set to `<destination-seed>` as well. Then the normal reconciliation flow will properly keep the correct `.spec.seedName` in the `BackupEntry`.

> inconveniently crashes just before this "Backup" step.

If it crashes before it moved the `BackupEntry` (which is more or less anyway just moving metadata) then it will retry the whole migration preparation flow again. If it crashes afterwards then it has already set the status to `Succeeded` and everything is OK.